File: backend/app/schemas/market_expectations.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

MarketExpectationInterval = Literal["1h", "4h", "1d", "1w"]


class ApiModel(BaseModel):
    model_config = ConfigDict(populate_by_name=True)


class MarketExpectationPoint(ApiModel):
    timestamp: datetime
    probability: float


class MarketExpectationSeries(ApiModel):
    candidate_catalog_id: int = Field(alias="candidateCatalogId")
    candidate_name: str = Field(alias="candidateName")
    display_name: str = Field(alias="displayName")
    market_id: str = Field(alias="marketId")
    points: list[MarketExpectationPoint]


class MarketExpectationsMetadata(ApiModel):
    latest_timestamp: datetime | None = Field(alias="latestTimestamp")


class MarketExpectationLeader(ApiModel):
    candidate_catalog_id: int = Field(alias="candidateCatalogId")
    display_name: str = Field(alias="displayName")
    probability: float


class MarketExpectationLeaderMargin(ApiModel):
    value: float
    leader_candidate_catalog_id: int = Field(alias="leaderCandidateCatalogId")
    runner_up_candidate_catalog_id: int = Field(alias="runnerUpCandidateCatalogId")


class MarketExpectationLargestChange(ApiModel):
    candidate_catalog_id: int = Field(alias="candidateCatalogId")
    display_name: str = Field(alias="displayName")
    value: float
    absolute_value: float = Field(alias="absoluteValue")


class MarketExpectationsSummary(ApiModel):
    current_leader: MarketExpectationLeader | None = Field(alias="currentLeader")
    leader_margin: MarketExpectationLeaderMargin | None = Field(alias="leaderMargin")
    largest_change: MarketExpectationLargestChange | None = Field(alias="largestChange")


class MarketExpectationsResponse(ApiModel):
    sources: list[str]
    metadata: MarketExpectationsMetadata
    summary: MarketExpectationsSummary
    series: list[MarketExpectationSeries]

File: frontend/src/models/marketExpectations.ts
export type MarketExpectationInterval = '1h' | '4h' | '1d' | '1w';

export type MarketExpectationDateRange = {
  max: string | null;
  min: string | null;
};

export type MarketExpectationOptionCandidate = {
  candidateCatalogId: number;
  displayName: string;
  latestProbability: number;
};

export type MarketExpectationOptions = {
  candidates: MarketExpectationOptionCandidate[];
  dateRange: MarketExpectationDateRange;
  defaultCandidateCatalogIds: number[];
  intervals: MarketExpectationInterval[];
};

export type MarketExpectationPoint = {
  probability: number;
  timestamp: string;
};

export type MarketExpectationSeries = {
  candidateCatalogId: number;
  candidateName: string;
  displayName: string;
  marketId: string;
  points: MarketExpectationPoint[];
};

export type MarketExpectationMetadata = {
  latestTimestamp: string | null;
};

export type MarketExpectationLeader = {
  candidateCatalogId: number;
  displayName: string;
  probability: number;
};

export type MarketExpectationLeaderMargin = {
  leaderCandidateCatalogId: number;
  runnerUpCandidateCatalogId: number;
  value: number;
};

export type MarketExpectationLargestChange = {
  absoluteValue: number;
  candidateCatalogId: number;
  displayName: string;
  value: number;
};

export type MarketExpectationSummary = {
  currentLeader: MarketExpectationLeader | null;
  largestChange: MarketExpectationLargestChange | null;
  leaderMargin: MarketExpectationLeaderMargin | null;
};
